### PR TITLE
mpi backend added

### PR DESCRIPTION
**Pull request by Akado2009**
Fixes: https://github.com/kubeflow/pytorch-operator/issues/66

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pytorch-operator/68)

### CODE REVIEW

**Comment by johnugeorge** [comment]
Can we add have consistent folder naming  similar to that of dist-mnist? 

mpi-mnist/cpu/Dockerfile
mpi-mnist/cpu/MPI_mnist_CPU.py
mpi-mnist/cpu/v1alpha2/mpi_mnist_job_cpu.yaml

similar for gpu 
mpi-mnist/gpu/..

File: examples/mpi-mnist/mpi_backend/cpu/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+     /opt/conda/bin/conda install -y -c pytorch magma-cuda90 && \
+     /opt/conda/bin/conda install -c conda-forge openmpi && \
+     /opt/conda/bin/conda clean -ya
+ENV PATH /opt/conda/bin:$PATH
+# This must be done before pip so that requirements.txt is available
+WORKDIR /opt/pytorch
+#COPY . .
+RUN git clone --recursive https://github.com/pytorch/pytorch
+RUN cd pytorch/
+RUN ls -lah
```
Comment (gaocegege) [inline examples/mpi-mnist/mpi_backend/cpu/Dockerfile]: Could we merge these run scripts into one to reduce the size of the docker image?

File: examples/mpi-mnist/mpi_backend/gpu/mpi_mnist_job_gpu.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+            - name: pytorch
+              image: docker.io/akado2009/pytorch-mpi-mnist-gpu
+    Worker:
+      replicas: 3
+      restartPolicy: OnFailure
+      template:
+        spec:
+          containers:
+            - name: pytorch
+              image: docker.io/akado2009/pytorch-mpi-mnist-gpu
```
Comment (gaocegege) [inline examples/mpi-mnist/mpi_backend/gpu/mpi_mnist_job_gpu.yaml]: Please add a new line at the end of the file.

File: examples/mpi-mnist/mpi_backend/cpu/Dockerfile (hunk 1 above)
Comment (Akado2009) [reply examples/mpi-mnist/mpi_backend/cpu/Dockerfile]: Those are unnecessary. Fixed.

File: examples/mpi-mnist/mpi_backend/gpu/mpi_mnist_job_gpu.yaml (hunk 2 above)
Comment (Akado2009) [reply examples/mpi-mnist/mpi_backend/gpu/mpi_mnist_job_gpu.yaml]: Fixed.

**Comment by Akado2009** [comment]
Working now on adding the tests for my example.

File: examples/mpi-mnist/cpu/v1alpha2/mpi_mnist_job_cpu.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+            - name: pytorch
+              image: docker.io/akado2009/pytorch-mpi-mnist-cpu
+    Worker:
+      replicas: 3
+      restartPolicy: OnFailure
+      template:
+        spec:
+          containers:
+            - name: pytorch
+              image: docker.io/akado2009/pytorch-mpi-mnist-cpu
```
Comment (johnugeorge) [inline examples/mpi-mnist/cpu/v1alpha2/mpi_mnist_job_cpu.yaml]: this has to be changed when the image is uploaded to official kubeflow registry (like other examples)

File: scripts/v1alpha2/run-tests-mpi-cpu.sh [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+KF_ENV=pytorch
+
+echo "Activating service-account"
+gcloud auth activate-service-account --key-file=${GOOGLE_APPLICATION_CREDENTIALS}
+echo "Configuring kubectl"
+gcloud --project ${PROJECT} container clusters get-credentials ${CLUSTER_NAME} \
+    --zone ${ZONE}
+
+cd ${APP_NAME}
+echo "Install PyTorch v1alpha2 operator"
```
Comment (johnugeorge) [inline scripts/v1alpha2/run-tests-mpi-cpu.sh]: verify if you need to install operator again ? It is already installed in run-defaults.sh.

File: examples/mpi-mnist/cpu/MPI_mnist_CPU.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python
```
Comment (johnugeorge) [inline examples/mpi-mnist/cpu/MPI_mnist_CPU.py]: File names - > lower_case_with_underscores

File: scripts/v1alpha2/run-tests-mpi-gpu.sh [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+KF_ENV=pytorch
+
+echo "Activating service-account"
+gcloud auth activate-service-account --key-file=${GOOGLE_APPLICATION_CREDENTIALS}
+echo "Configuring kubectl"
+gcloud --project ${PROJECT} container clusters get-credentials ${CLUSTER_NAME} \
+    --zone ${ZONE}
+
+cd ${APP_NAME}
+echo "Install PyTorch v1alpha2 operator"
```
Comment (johnugeorge) [inline scripts/v1alpha2/run-tests-mpi-gpu.sh]: same

File: scripts/v1alpha2/run-cleanpodpolicy-all.sh [hunk 7]
```diff
@@ -50,3 +50,12 @@ echo "Running mnist test"
 MNIST_TEST_IMAGE_TAG="pytorch-dist-mnist_test:1.0"
 go run ./test/e2e/v1alpha2/cleanpolicy_all.go --namespace=${NAMESPACE} --image=${REGISTRY}/${MNIST_TEST_IMAGE_TAG} --name=mnistjob-cleanall
```
Comment (johnugeorge) [inline scripts/v1alpha2/run-cleanpodpolicy-all.sh]: do we need to test cleanpod policy with mpi images also?  I feel it is not needed

File: scripts/build.sh [hunk 8]
```diff
@@ -49,3 +49,13 @@ gcloud container builds submit ./examples/dist-sendrecv/ --tag=${REGISTRY}/${SEN
 echo "Building MNIST test image"
 MNIST_TEST_IMAGE_TAG="pytorch-dist-mnist_test:1.0"
 gcloud container builds submit ./examples/dist-mnist/ --tag=${REGISTRY}/${MNIST_TEST_IMAGE_TAG} --project=${PROJECT}
+
+echo "Building MPI CPU test image"
+REGISTRY="docker.io"
+CPU_TEST_IMAGE="akado2009/pytorch-mpi-mnist-cpu"
```
Comment (johnugeorge) [inline scripts/build.sh]: use official registry itself and test image as pytorch-mpi-mnist-cpu:1.0

File: scripts/build.sh [hunk 9]
```diff
@@ -49,3 +49,13 @@ gcloud container builds submit ./examples/dist-sendrecv/ --tag=${REGISTRY}/${SEN
...
+gcloud container builds submit ./examples/mpi-mnist/cpu --tag=${REGISTRY}/${CPU_TEST_IMAGE} --project=${PROJECT}
+
+echo "Building MPI GPU test image"
+GPU_TEST_IMAGE="akado2009/pytorch-mpi-mnist-gpu"
```
Comment (johnugeorge) [inline scripts/build.sh]: same here

**Pull request closed without merging**